### PR TITLE
[TACHYON-973] Clearly separate Client implmentation and interface

### PR DESCRIPTION
**Pull request by calvinjia**
https://tachyon.atlassian.net/browse/TACHYON-973

### CODE REVIEW

File: clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+  /**
+   * Constructor, currently TachyonFileSystem does not retain any state
+   */
+  protected AbstractTachyonFileSystem() {
+    mContext = FileSystemContext.INSTANCE;
+  }
+
+  /**
+   * {@inheritDoc}
```
Comment (yupeng9) [inline clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java]: Question on convention: do you need it? I think the override annotation might be enough.

Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java]: Yeah for the ones without additional documentation I think I can remove the `{@inheritDoc}`

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystem.java [hunk 2]
```diff
@@ -168,132 +118,37 @@ public FileInStream getInStream(TachyonFile file, ClientOptions options) throws
...
-      mContext.releaseMasterClient(masterClient);
-    }
-  }
-
-  /**
-   * Adds metadata about a file in the under storage system to Tachyon. Only metadata will be
-   * updated and no data will be transferred. The data can be added to Tachyon space by doing an
-   * operation with the cache option specified, for example reading.
+   * Alternative way to get a FileOutStream to a file that has already been created. This should
+   * not be used and will be removed in a later release.
```
Comment (yupeng9) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystem.java]: Can you add the version when this method is deprecated?

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java [hunk 3]
```diff
@@ -30,29 +31,48 @@
  * including folders. Clients should provide their own interface for reading/writing files.
  */
 @PublicApi
-interface TachyonFSCore {
+interface TachyonFileSystemCore {
```
Comment (yupeng9) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java]: Do we need the 'core' suffix?

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java [hunk 4]
```diff
@@ -30,29 +31,48 @@
...
 
   /**
-   * Deletes a file. If the file is a folder, its contents will be deleted recursively.
+   * Creates a file with the provided blocksize as the standard blocksize of the file. If the file's
```
Comment (yupeng9) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java]: block size

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java [hunk 5]
```diff
@@ -30,29 +31,48 @@
...
+   * parent directories do not exist, they will be created if the recursive flag is set.
+   *
+   * @param path the path of the file to create in Tachyon space
+   * @param blockSize the block size in bytes, must be greater than 0
+   * @param recursive whether or not to create parent directories if required
+   * @return the file id that identifies the newly created file
+   * @throws BlockInfoException if the block size is less than 0
```
Comment (yupeng9) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java]: less and equal?

**Comment by yupeng9** [comment]
Thanks for the improvement on the client hierarchy. I left some comments.

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java (hunk 3 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java]: I wanted to call the implementation `TachyonFileSystem`

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java (hunk 5 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java]: good catch, updated.

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java (hunk 4 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystemCore.java]: updated.

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystem.java (hunk 2 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystem.java]: sure, added.

**Comment by yupeng9** [comment]
LGTM

File: clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+      mContext.releaseMasterClient(masterClient);
+    }
+  }
+
+  @Override
+  public boolean mkdirs(TachyonURI path, boolean recursive) throws InvalidPathException,
+      IOException, FileAlreadyExistException {
+    FileSystemMasterClient masterClient = mContext.acquireMasterClient();
+    try {
+      // TODO: Change this RPC's arguments
```
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java]: change this RPC arguments to what?

File: clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+      boolean result = masterClient.createDirectory(path.getPath(), true);
```
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java]: should the second argument be 'recursive'?

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystem.java [hunk 8]
```diff
@@ -37,97 +31,57 @@
...
- * input/output streams.
+ * A TachyonFileSystem implementation including convenience methods as well as a streaming API to
+ * read and write files. Creating files is not supported by this TachyonFileSystem because the
+ * files should only be written once, thus getOutStream is sufficient for creating and writing to
+ * a file.
  */
 @PublicApi
-public final class TachyonFileSystem implements Closeable, TachyonFSCore {
-  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+public class TachyonFileSystem extends AbstractTachyonFileSystem {
```
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystem.java]: Is this class NOT supposed to call the master, and all the master calls should happen in the AbstractTachyonFileSystem?

**Comment by gpang** [comment]
I just left a few comments. Thanks!

File: clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java (hunk 7 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java]: good catch, thanks.

File: clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystem.java (hunk 8 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystem.java]: Yeah, anything it implements should use the methods in AbstractTachyonFileSystem.

File: clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java (hunk 6 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/AbstractTachyonFileSystem.java]: I think this is outdated and can be removed.

**Comment by gpang** [comment]
LGTM

**Pull request merged**